programs/loader-v3/src/lib.rs

```diff
@@ -0,0 +1,1369 @@
+                    );
+                    return Err(InstructionError::InvalidArgument);
+                }
+                if state.is_deployed {
+                    ic_logger_msg!(log_collector, "Destination program is not retracted");
+                    return Err(InstructionError::InvalidArgument);
+                }
+                let buffer = if let Some(ref source_program) = source_program {
+                    let source_state = check_program_account(
+                        &log_collector,
+                        instruction_context,
+                        source_program,
+                        authority_address,
+                    )?;
+                    if source_state.is_deployed {
+                        ic_logger_msg!(log_collector, "Source program is not retracted");
+                        return Err(InstructionError::InvalidArgument);
+                    }
+                    source_program
+                } else {
+                    &program
+                };
+                let (_executor, load_program_metrics) = load_program_from_account(
+                    &invoke_context.feature_set,
+                    invoke_context.get_compute_budget(),
+                    invoke_context.get_log_collector(),
+                    buffer,
+                    use_jit,
+                    false,
+                )?;
+                load_program_metrics.submit_datapoint(&mut invoke_context.timings);
+                if let Some(mut source_program) = source_program {
+                    let rent = invoke_context.get_sysvar_cache().get_rent()?;
+                    let required_lamports = rent.minimum_balance(program.get_data().len());
+                    let transfer_lamports =
+                        program.get_lamports().saturating_sub(required_lamports);
+                    program.set_data_from_slice(source_program.get_data())?;
+                    source_program.set_data_length(0)?;
+                    source_program.checked_sub_lamports(transfer_lamports)?;
+                    program.checked_add_lamports(transfer_lamports)?;
+                }
+                let state = get_state_mut(program.get_data_mut()?)?;
+                state.slot = current_slot;
+                state.is_deployed = true;
+            }
+            LoaderV3Instruction::Retract => {
+                let mut program =
+                    instruction_context.try_borrow_instruction_account(transaction_context, 0)?;
+                let authority_address = transaction_context.get_key_of_account_at_index(
+                    instruction_context.get_index_of_instruction_account_in_transaction(1)?,
+                )?;
+                let state = check_program_account(
+                    &log_collector,
+                    instruction_context,
+                    &program,
+                    authority_address,
+                )?;
+                if !state.is_deployed {
```

Comment:
Should we check `DEPLOYMENT_COOLDOWN_IN_SLOTS` for the program before it can be retracted?

Reply:
That would prevent the user from upgrading a program in a single transaction.

Reply:
I was mostly viewing it by the comment here: https://github.com/solana-labs/solana/pull/30464/files#diff-122620648d15121d8a722c52f2dcf42e3587ad87b0c254162edc8f857d6d218fR9

```
/// Cooldown before a program can be un-/redeployed again
pub const DEPLOYMENT_COOLDOWN_IN_SLOTS: u64 = 750;
```

I am likely confused.

Reply:
No, you are right, I misunderstood the first comment. Yes, we should check it here as well, just not reset / restart the cooldown.